### PR TITLE
Multiple virtualization fixes

### PR DESCRIPTION
**Pull request by grisu48**
Multiple virtualization fixes after last round of XEN runs.

Related issues:
- [Fix failing xen test runs due to increased number of workers](https://progress.opensuse.org/issues/93576)
- [Fix Windows test run on xen](https://progress.opensuse.org/issues/93435)

Verification run: http://openqa.qam.suse.cz/tests/23166#dependencies

### CODE REVIEW

**Comment by grisu48** [comment]
Could be merged if the VR passes.

**Review by pdostal — APPROVED** [review]

File: tests/virtualization/universal/windows.pm [hunk 1]
```diff
@@ -45,7 +49,14 @@ sub run {
...
-    remove_guest $_ foreach (keys %virt_autotest::common::imports);
+    remove_guests();
+}
+
+
+sub post_fail_hook {
+    my $self = shift;
+    remove_guests();
+    $self->SUPER::post_fail_hook;
 }
```
Comment (pdostal) [inline tests/virtualization/universal/windows.pm:60]: No, please don't do this. If this module fails we won't have VM to debug.

File: tests/virtualization/universal/windows.pm [hunk 2]
```diff
@@ -26,11 +26,15 @@ sub remove_guest {
     }
 }
 
+sub remove_guests() {
+    remove_guest $_ foreach (keys %virt_autotest::common::imports);
+}
```
Comment (pdostal) [inline tests/virtualization/universal/windows.pm:31]: Two things:
 * Why is this here and not in `virt_autotest::utils`?
 * Whys this even exists? We have ton of functions nobody barely uses.

**Review by pdostal — CHANGES_REQUESTED** [review]

File: tests/virtualization/universal/windows.pm (hunk 1 above)
Comment (grisu48) [reply tests/virtualization/universal/windows.pm]: Good point. I left the `post_fail_hook` with a note in there, so that we know that we do this on purpose from now on.

File: tests/virtualization/universal/windows.pm (hunk 2 above)
Comment (grisu48) [reply tests/virtualization/universal/windows.pm:31]: I was using it twice within the same module and line 37 looks much slimmer with this.
IMHO this function is really too small to justify a lib function.

File: tests/virtualization/universal/windows.pm [hunk 3]
```diff
@@ -44,8 +49,18 @@ sub run {
     assert_script_run "ssh $username\@$_ 'systeminfo' | tee /tmp/$_-systeminfo.txt"                            foreach (keys %virt_autotest::common::imports);
     upload_logs "/tmp/$_-systeminfo.txt"                                                                       foreach (keys %virt_autotest::common::imports);
     assert_script_run "ssh $username\@$_ 'systeminfo' | grep '$virt_autotest::common::imports{$_}->{version}'" foreach (keys %virt_autotest::common::imports);
+}
 
-    remove_guest $_ foreach (keys %virt_autotest::common::imports);
+sub post_fail_hook {
+    my $self = shift;
+    # Note: Don't cleanup guests on test failure, so their state is preserved for debugging purposes!
```
Comment (grisu48) [inline tests/virtualization/universal/windows.pm:56]: Note: This looks not necessary but I'd leave it here so that we know in the future, that one should not do a cleanup on test failure. So: No functional advantage but conservation of thought.

File: tests/virtualization/universal/hotplugging.pm [hunk 4]
```diff
@@ -216,6 +216,11 @@ sub run_test {
         save_screenshot;
     }
 
+    # 0. Guest preparation - Increase maximum memory for this test run
+    shutdown_guests();
+    assert_script_run("virsh setmaxmem $_ 4G --config && virsh start $_") foreach (keys %virt_autotest::common::guests);
```
Comment (alice-suse) [inline tests/virtualization/universal/hotplugging.pm:221]: How about the guests with mem >=4G already?

Comment (grisu48) [reply tests/virtualization/universal/hotplugging.pm:221]: We don't have any SLES guests with >= 4G at the moment and given the memory constrains I doubt we will have them in the near future, so I fail to see the problem you'd like to solve.

Could you illustrate the problem you want to solve? Thinking about adding >4GB guests in the future or thinking about testing >4GB?

Comment (alice-suse) [reply tests/virtualization/universal/hotplugging.pm:221]: This is not a problem for now definitely. But there will likely be such bigger mem guests in not far future. So suggest to make maxmem configurable from testsuite level, if not much effort needed.

File: tests/virtualization/universal/windows.pm [hunk 5]
```diff
@@ -26,11 +26,16 @@ sub remove_guest {
     }
 }
 
+sub remove_guests() {
```
Comment (pdostal) [inline tests/virtualization/universal/windows.pm:29]: At least please make this more descriptive.
I don't understand the purpose of one line function tho.

File: lib/virt_autotest/utils.pm [hunk 6]
```diff
@@ -393,9 +402,12 @@ sub wait_guests_shutdown {
 
 # Start all guests and wait until they are online
 sub start_guests {
-    script_run "virsh start $_" foreach (keys %virt_autotest::common::guests);
-    # Wait for guests to show up as running
-    script_retry("virsh list | grep $_ | grep running", delay => 5, retry => 10) foreach (keys %virt_autotest::common::guests);
+    foreach my $guest (keys %virt_autotest::common::guests) {
+        if (script_run("virsh list | grep '$guest' | grep running") != 0) {
+            assert_script_run("virsh start '$guest'");
+            script_retry("virsh list | grep '$guest' | grep running", delay => 5, retry => 10);
```
[the remark this answers is not in the document]
Comment (grisu48) [reply lib/virt_autotest/utils.pm:408]: For booting it is for sure, but this is about starting the VM. You think just starting the VM (not booting the system!) we need 2.5 minutes?

File: lib/main_common.pm [hunk 7]
```diff
@@ -2659,6 +2659,7 @@ sub load_virtualization_tests {
 sub load_hypervisor_tests {
     return unless (get_var('HOST_HYPERVISOR') =~ /xen|kvm|qemu/);
     return unless get_var('VIRT_PART');
+    my $windows = check_var('VIRT_PART', 'windows');
```
[the remark this answers is not in the document]
Comment (grisu48) [reply lib/main_common.pm:2662]: It is used twice: line 2692 and line 2755, see https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/12647/commits/ace922d5331bae770f9bc6f4ea9a26131b77555b#diff-051cb1f26a96a4f724250d9182535d868c691e20963ef91a1a7ca109e9538379R2692

File: tests/virtualization/universal/hotplugging.pm (hunk 4 above)
Comment (grisu48) [reply tests/virtualization/universal/hotplugging.pm:221]: Ok there is a problem with making this configurable: We have [too many xen guests](https://progress.opensuse.org/issues/93576) at the moment, this is why we had to limit the maximum memory per guest from 4GB to 2 GB. In hotplugging however we increase the amount of memory to 4GB otherwise we cannot test the memory increase.

I'm gonna tackle this issue by increasing the maximum memory in `hotplugging` by 2GB additional to the maximum memory. This should give us the freedom to configure the maximum memory, while it also allows us to keep our guests with 2GB of ram.

File: tests/virtualization/universal/windows.pm (hunk 5 above)
Comment (grisu48) [reply tests/virtualization/universal/windows.pm:29]: Ad your first comment: `virsh setmaxmem` needs to run before the guest is started, at least for KVM, that's why it is there. I will move it however to a more descriptive subroutine, this should make it easier to understand.

Ad second comment: ok I give in and remove my function.

File: tests/virtualization/universal/hotplugging.pm (hunk 4 above)
Comment (grisu48) [reply tests/virtualization/universal/hotplugging.pm]: @alice-suse Please review my current suggestion and let me know, if this is what you had in mind.

File: lib/virt_autotest/utils.pm (hunk 6 above)
[the remark this answers is not in the document]
Comment (pdostal) [reply lib/virt_autotest/utils.pm:408]: Ah true. OK, keep it :)

**Review by pdostal — APPROVED** [review]
All good from my side 😄

**Comment by grisu48** [comment]
I've marked it as draft, as we need to re-run the VR with the current changes.

**Review by alice-suse — APPROVED** [review]

File: tests/virtualization/universal/hotplugging.pm (hunk 4 above)
Comment (grisu48) [reply tests/virtualization/universal/hotplugging.pm]: I assume the thumbsup means your comment has been addressed. I'm marking this as resolved, feel free to reopen if there are still concerns.
